### PR TITLE
docs: update contributing guide

### PR DESCRIPTION
**Pull request by tilleryd**
@kaitmore @shanberg I was trying to think of more examples of what not to do or do. Please feel free to suggest more, otherwise the two should suffice.

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -1,5 +1,12 @@
 # Contributing
 
+## Guidelines
+
+The intent of this library is to provide apps with a set of re-usable components that have purpose and meaning. We do not seek to create dopelgangers of HTML primitives like `<p/>`, `<h1/>`, `<a/>`, etc. Therefore, when proposing or creating new components consider these things:
+
+1. Can you achieve the same result by styling an HTML primitive? If so, do not add it to this library.
```
Comment (shanberg) [inline CONTRIBUTING.md]: I'm not sure about this. I can see us creating a `<Heading />` or `<ThingHeading/>` component which only styles an `h1/h2/h3`, but applies appropriate margins and relationships to sibling elements. Many styled headings, links, or buttons would just be single elements, but specially styled. Simpler resulting markup is a benefit, not a problem.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -1,5 +1,12 @@
...
+2. Does this component add significant features by adding behavior to an HTML primitive? For example, adding a tooltip to a `<p/>` tag is actually a Tooltip component, not an extenstion of a `<p/>` tag and would thus warrant a new Tooltip component.
+
```
Comment (shanberg) [inline CONTRIBUTING.md]: How about...

Components should serve a single clear use case: "Shell panel headings", "App navigation tabs", "Context menu action button". If you find yourself frequently having to provide style overrides, that's a sign the component is not well-targeted to where it’s being used. If you find yourself frequently having to provide the same style overrides over and over to a component, that's a sign you should extend the component to better target its use context.

File: CONTRIBUTING.md (hunk 1 above)
Comment (tilleryd) [reply CONTRIBUTING.md]: Can you give me an example of where a heading tag would style sibling elements? I get what you're saying. I'm just trying to figure out if this is a contrived example for the sake of discussion.

File: CONTRIBUTING.md (hunk 2 above)
Comment (tilleryd) [reply CONTRIBUTING.md]: I defer to your eloquence...

File: CONTRIBUTING.md (hunk 1 above)
Comment (shanberg) [reply CONTRIBUTING.md]: I was repeating myself, not actually adding a second point. Styles would affect a `<Heading>`'s _relationship to_ sibling elements, not the sibling elements themselves. The margin is a relationship to a sibling. Paddings, borders, backgrounds, shadows, alignment, and display properties also have some kind of relationship aspect.

There may be edge cases where it’s helpful to be able to style a sibling, but those are usually complicated and best avoided.

Comment (tilleryd) [reply CONTRIBUTING.md]: Ok good, I agree regarding influence on surrounding elements.

Comment (tilleryd) [reply CONTRIBUTING.md]: I think this comes down to naming. I've updated the text.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -1,5 +1,12 @@
 # Contributing
 
+## Guidelines
+
+The intent of this library is to provide apps with a set of re-usable components that have a beneficial, re-usable purpose. When creating or updating components, please keep these guidelines in mind.
+
+1. Semantic naming is important and should clearly communicate the components purpose. When naming a component, choose one that's contextual to its use. For example, a component built around an `h1` element should be named `Heading` rather than `H1`. This has the added benefit of clearly separating custom components with primitive HTML elements.
```
Comment (shanberg) [inline CONTRIBUTING.md]: 1. Component naming should communicate the component's purpose, not its construction. For example, a _layout section heading_ built around an `<h1/>` element should be named `LayoutSectionHeading`, not `H1`.

**Comment by tilleryd** [comment]
> Nice, this is great!

I humbly bow to @shanberg for his eloquent words.

**Pull request merged**